### PR TITLE
Add type signatures to `lru_cache` using overloads (not ready to merge)

### PR DESCRIPTION
**Pull request by alwaysmpe**
Undecided if this is a good way to go, but marked as for review to see what others say. I've added some notes on output of `mypy_primer` at the end, there are a couple of unexpected errors but mostly good.

Change `lru_cache` callable signature to use `ParamSpec`. This has previously been tried unsuccessfully (#11662) due to compatibility with the `classmethod` decorator. I've attempted to solve that compatibility by using overloads that do/don't include the necessary `cls` and `self` parameter when it appearrs that a method/classmethod is being cached.

This removes the `Hashable` constraint on parameters.

Instead, this uses the function signatures (almost) correctly (methods and class methods have 2 overloads, one with and without `cls` and `self`, but I think this would be less error prone than currently, could possibly be refined).

As a trivial example, the below type checks correctly
```python
from functools import cache
@cache
def foo(arg: str):
    ...
foo(1) # correctly raises type error
foo("1")
foo.cache_clear()
```
There is a more comprehensive example that I've been using to test in #12952 which (with 2 minor exceptions) type checks perfectly.

### CODE REVIEW

**Comment by brianschubert** [comment]
(I'm guessing this follows from #12952 for context)

**Comment by alwaysmpe** [comment]
> (I'm guessing this follows from #12952 for context)

Yup, this is (another) attempt, but quite a bit simpler than the previous approach, it feels like it's almost doable (and this caused me a few bugs) so I'm determined.

**Comment by Daverball** [comment]
This solution appears to depend on methods/classmethods following the convention of naming the first parameter `self`/`cls` respectively, so I feel like this will break as soon as you violate this convention or have a positional only `self`/`cls` because one or more of your other arguments are positional only. Can you add some test cases to verify that this solution still works correctly with positional only `self`/`cls` or when you break this naming convention?

As soon as you stop relying on the parameter name you can no longer write a `Protocol` that properly distinguishes between a regular method and a classmethod (you would have to rely on the type checker always preferring the `type[_T]` overload for `classmethod` and always preferring `_T` for regular methods, which does not seem possible with metaclasses), and that's usually where earlier attempts failed.

File: stdlib/functools.pyi [hunk 1]
```diff
@@ -51,22 +53,46 @@ if sys.version_info >= (3, 9):
...
         typed: bool
 
+class _Function(Protocol[_P, _R]):  # type: ignore[misc]  # pyright: ignore reportInvalidTypeVarUse
+    def __call__(self, *args: _P.args, **kwds: _P.kwargs) -> _R: ...
+
+class _Method(Protocol[_T, _P, _R]):  # type: ignore[misc]  # pyright: ignore reportInvalidTypeVarUse
+    def __call__(_self, self: _T, *args: _P.args, **kwds: _P.kwargs) -> _R: ...
+
+class _Classmethod(Protocol[_T, _P, _R]):  # type: ignore[misc]  # pyright: ignore reportInvalidTypeVarUse
+    def __call__(_self, cls: type[_T], *args: _P.args, **kwds: _P.kwargs) -> _R: ...
```
Comment (alwaysmpe) [inline stdlib/functools.pyi:63]: I think combine these and use `Concatenate` instead, avoid needing to explicitly name a parameter, fix issue mentioned by @Daverball

**Comment by alwaysmpe** [comment]
> you would have to rely on the type checker always preferring the `type[_T]` overload for `classmethod` and always preferring `_T` for regular methods

I'm hoping that shouldn't be a problem because overloads are only usable if it's possible to substitute in all parameters and if `self` generic argument isn't a `Callable[[type[T]], Any]` then the `classmethod` overload isn't selectable, but I'm putting together some tests, probably not until after work.

Thankyou for ther feedback.

File: stdlib/functools.pyi (hunk 1 above)
Comment (Daverball) [reply stdlib/functools.pyi:63]: I think that's pretty much what the closest to working earlier approach did, yes. But it still had its edge cases where it didn't work unfortunately. Maybe things are a little bit more forgiving now.

If I'm not mistaken what you're doing in addition to earlier approaches is to always accept both signatures for methods/classmethods (i.e. the one where you have to explicitly provide `self`/`cls` and the one where you don't), so you give up some type safety on `instance.method()` vs. `Cls.method()` since it would not be able to detect that you're missing a required parameter.

That might be a reasonable trade-off for the overall better type safety, although I fear that language servers will not know what to do with these overloads, so instead of no parameter hints you may get confusing/difficult to read hints, which could be seen as a regression in ergonomics.

**Comment by Daverball** [comment]
> I'm hoping that shouldn't be a problem because overloads are only usable if it's possible to substitute in all parameters and if `self` generic argument isn't a `Callable[[type[T]], Any]` then the `classmethod` overload isn't selectable, but I'm putting together some tests, probably not until after work.

That wouldn't be the case for metaclasses or instances of classes that define a `__call__` decorated with `classmethod`, so there are edge cases here unfortunately. Also the overload resolution can always select the regular method case over the classmethod one once the named parameter stops helping you differentiate, they will always overlap for classmethods.

It might be worth just reducing this to the core of your idea and use `Concatenate` to always split off the first parameter and then accept calls with or without it. It seems like that might always work, albeit not fully type safe.

**Comment by alwaysmpe** [comment]
Tests are going to fail. There are some things that are there to show what is/isn't working, pyright ignores are correct and most of the typing behavior is as intended, there are a couple of false positives/negatives. I haven't got all of the mypy ignores correct.

I've also been doing some work to refine the descriptor approach on [another branch](https://github.com/alwaysmpe/typeshed/tree/cache-descriptor), that is probably more robust from a typing perspective. Will create another PR to run primer for comparison (apologies for the PR noise, it's all well intentioned if a bit rushed sometimes).

**Comment by alwaysmpe** [comment]
Closing this. The correct approach is for builtin descriptor type definitions to match their runtime descriptor protocol behaviour but `mypy` (and I believe other type checkers) don't currently understand it correctly. So I'll try to fix that first. If interested I discuss this on [typing forum](https://discuss.python.org/t/why-do-some-all-type-checkers-discard-classmethod-staticmethod-property-instead-of-using-the-descriptor-protocol/72261)
